src/Data/Store.elm

```diff
@@ -0,0 +1,233 @@
+    )
+
+
+buildFoldersDataAndRelations : List AudienceFolder -> ( Data, Relations ) -> ( Data, Relations )
+buildFoldersDataAndRelations listOfFolders acc =
+    List.foldr
+        (\folder ( data, relations ) ->
+            ( { data | folders = Dict.insert folder.id folder data.folders }
+            , case folder.parent of
+                Nothing ->
+                    { relations | root = insertFolderToLevel folder.id relations.root }
+
+                Just parentID ->
+                    let
+                        nextLevel =
+                            Dict.get parentID relations.levels
+                                |> Maybe.withDefault emptyLevel
+                                |> insertFolderToLevel folder.id
+                    in
+                    { relations | levels = Dict.insert parentID nextLevel relations.levels }
+            )
+        )
+        acc
+        listOfFolders
+
+
+buildAudiencesDataAndRelations : List Audience -> ( Data, Relations ) -> ( Data, Relations )
+buildAudiencesDataAndRelations listOfAudiences acc =
+    List.foldr
+        (\audience ( data, relations ) ->
+            ( { data | audiences = Dict.insert audience.id audience data.audiences }
+            , case audience.folder of
+                Nothing ->
+                    let
+                        ( nextShared, nextRoot ) =
+                            insertAudienceToLevel audience relations.shared relations.root
+                    in
+                    { relations
+                        | root = nextRoot
+                        , shared = nextShared
+                    }
+
+                Just folderID ->
+                    let
+                        ( nextShared, nextLevel ) =
+                            Dict.get folderID relations.levels
+                                |> Maybe.withDefault emptyLevel
+                                |> insertAudienceToLevel audience relations.shared
+                    in
+                    { relations
+                        | levels = Dict.insert folderID nextLevel relations.levels
+                        , shared = nextShared
+                    }
+            )
+        )
+        acc
+        listOfAudiences
+
+
+init : List AudienceFolder -> List Audience -> Store
+init listOfFolders listOfAudiences =
+    let
+        ( data, relations ) =
+            ( initialData, initialRelations )
+                |> initEmptyLevels listOfFolders
+                |> buildFoldersDataAndRelations listOfFolders
+                |> buildAudiencesDataAndRelations listOfAudiences
+    in
+    Store data relations
+
+
+extractListFromData : List comparable -> Dict comparable { a | id : comparable } -> List { a | id : comparable }
+extractListFromData ids entities =
+    List.filterMap (\id -> Dict.get id entities) ids
+
+
+type Selector
+    = OnlyShared
+    | OnlyCurated
+    | OnlyCuratedIn AudienceFolderID
```

Comment:
I slightly don't like how the "what we're displaying" (shared/curated/authored) is coupled to "where we're in the tree" (root/non-root folder). But I don't have any specific suggestions how to do this differently with this design..

Reply:
agree. not a big deal though